component/loki/source/aws_firehose/component.go

```diff
@@ -0,0 +1,171 @@
+	"github.com/grafana/agent/pkg/util"
+)
+
+func init() {
+	component.Register(component.Registration{
+		Name: "loki.source.awsfirehose",
+		Args: Arguments{},
+
+		Build: func(opts component.Options, args component.Arguments) (component.Component, error) {
+			return New(opts, args.(Arguments))
+		},
+	})
+}
+
+type Arguments struct {
+	Server       *fnet.ServerConfig  `river:",squash"`
+	ForwardTo    []loki.LogsReceiver `river:"forward_to,attr"`
+	RelabelRules flow_relabel.Rules  `river:"relabel_rules,attr,optional"`
+}
+
+func (a *Arguments) UnmarshalRiver(f func(v interface{}) error) error {
+	*a = Arguments{}
+	type arguments Arguments
+	err := f((*arguments)(a))
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Component is the main type for the `loki.source.awsfirehose` component.
+type Component struct {
+	// mut controls concurrent access to fanout
+	mut    sync.RWMutex
+	fanout []loki.LogsReceiver
+
+	// destination is the main destination where the TargetServer writes received log entries to
+	destination loki.LogsReceiver
+	rbs         []*relabel.Config
+
+	server *fnet.TargetServer
+
+	opts component.Options
+	args Arguments
+
+	// utils
+	serverMetrics *util.UncheckedCollector
+	logger        log.Logger
+}
+
+// New creates a new Component.
+func New(o component.Options, args Arguments) (*Component, error) {
+	c := &Component{
+		opts:          o,
+		destination:   make(loki.LogsReceiver),
+		fanout:        args.ForwardTo,
+		serverMetrics: util.NewUncheckedCollector(nil),
+
+		logger: log.With(o.Logger, "component", "aws_firehose_logs"),
+	}
+
+	o.Registerer.MustRegister(c.serverMetrics)
+
+	if err := c.Update(args); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Run starts a routine forwards received message to each destination component.
+func (c *Component) Run(ctx context.Context) error {
+	defer func() {
+		c.mut.Lock()
+		defer c.mut.Unlock()
+		c.shutdownServer()
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case entry := <-c.destination:
+			c.mut.RLock()
+			for _, receiver := range c.fanout {
+				receiver <- entry
+			}
+			c.mut.RUnlock()
+		}
+	}
+}
+
+// Update updates the component with a new configuration, restarting the server if needed.
+func (c *Component) Update(args component.Arguments) error {
+	var err error
+	c.mut.Lock()
+	defer c.mut.Unlock()
+
+	newArgs := args.(Arguments)
+	c.fanout = newArgs.ForwardTo
+
+	// todo(pablo): is it a good practice to keep a reference to the arguments in the
+	// component struct, used for comparing here rather than destructuring them?
+	if newArgs.RelabelRules != nil && len(newArgs.RelabelRules) > 0 {
+		c.rbs = flow_relabel.ComponentToPromRelabelConfigs(newArgs.RelabelRules)
+	}
+
+	serverNeedsUpdate := !reflect.DeepEqual(c.args.Server, newArgs.Server)
+	if !serverNeedsUpdate {
+		c.args = newArgs
+		return nil
+	}
+
+	c.shutdownServer()
+
+	jobName := strings.Replace(c.opts.ID, ".", "_", -1)
```

Comment:
Why do we need this, does the jobName end up at a label? (I also saw we use in the gcplog component, I'm just not sure we do need it.)

Reply:
In both gcp logs and here the component name with that replace, for example `loki_souce_awsfirehose` is used as prefix for all weaveworks server metrics